### DISCUSSION TITLE
Question re package naming (i.e. pydantic1 / pydantic2)

### DISCUSSION THREAD

**Discussion by tlambert03** [pydantic/pydantic#5402] [Pydantic V2]
First off, thanks a million for all of your hard work on pydantic v1 and the exciting changes in v2.  I'm thrilled by the prospects and have been diving in to begin migrating all of my pydantic-using packages.  Congrats on all the progress so far 🏆

As I see just how big of an API change it is, i find myself wondering how it's going to affect environment-solving going forward, and I'm curious whether there's already been a discussion about having a differently named package for v1 and v2.   I can definitely understand that you might not want to name the new version `pydantic2`, but seeing as it is mostly a different API, have you considered releasing a new `pydantic1` package on pypi (something that would not receive any updates after the final pydantic v1 release)?  (apologies if I've missed a pre-existing issue on this)

My primary concern is that there will be some packages that use pydantic for whom it's simply not worth the time to update, and they will likely leave `pydantic<2` in their install requires and be done with it (at least for the foreseable future).  Their package works fine for their needs and they're not looking to update.   This of course works fine for applications who aren't likely to appear in an environment with other packages (e.g. django or fastAPI apps, etc...).  But for libraries that try to coexist in python environment (where we obviously have a flat single-version package structure unlike node) I do fear a little bit the upcoming version clashes.

I do realize that the `pydantic1` strategy still won't work for packages that don't change their install requirements to `pydantic1` in their project metadata _and_ go through and update all of their imports... but it still seems like a possible escape hatch that will enable coexistence of packages that might otherwise break each other.

Please don't get me wrong: it's super clear you've been very transparent and diligent about all the breaking API changes.  Now is definitely the right time, and you're giving everyone plenty of time!  Curious to hear your thoughts or prior discussion on this.  thanks!

**Comment by tlambert03** [comment]
maybe to put this another way.  Do you imagine that it will be possible for packages that use pydantic <v2 and packages that use ≥v2 to co-exist in the same environment?  Or that that will be more or less impossible?  And if the latter, do you have thoughts on how to "help" the ecosystem deal with this incompatibility?  (such as a differently named package?)

**Comment by caniko** [comment]
This has been discussed before: https://github.com/pydantic/pydantic/discussions/4649

**Comment by tlambert03** [comment]
Thank you for your response.  I'm not sure I agree that's the same issue really.  That issue was really more about a single large project incrementally updating itself.  My point is about compatibility across multiple unrelated libraries with different authors: As soon as `python_library_A` updates itself and adds `pydantic>=2.0` to its pip install requirements, it will break `python_library_B` (that uses `pydantic<2.0`) in the same environment.

To summarize points that were made there, you said:

> It is impossible to install two separate versions of the same package in one environment; maybe possible with some dirty tricks (make a 2nd package with the old version). This seems more like a deployment issue, and not pydantic-related 🤔

I do, however, see this is a pydantic issue.  Obviously, pydantic is a very widely used package these days.  And while a major use case might be applications/projects that don't need to co-exist in an environment with other projects, another very valid use case is python *[libraries](https://iscinumpy.dev/post/app-vs-library/)* that use pydantic.  As soon as pydantic v2 is released as a non-pre-release on pypi under the name `pydantic`, any environment that has an updated package (with `pydantic>=2.0`) and an un-updated package (with `pydantic<2.0`) will immediately break (I'm referring to the kind of concerns that were clarified nicely in ["Should You Use Upper Bound Version Constraints?](https://iscinumpy.dev/post/bound-version-constraints/) ).  It's not a deployment issue *per se* (though I can see that an individual application could deal with this on the deployment side), it's an issue that every python package needs to consider when releasing changes that change the entire API.

I'm hoping that pydantic has a solution "ready" for libraries (not just applications) before releasing pydantic v2 under the name `pydantic` on PyPI.  Releasing either the current version as `pydantic1` or the new version as `pydantic2` would solve this issue (as this is the best we can do in python to allow two separate versions of the same package in one environment).  But just releasing pydantic v2 would leave a lot of communities with a massive headache

**Reply by dmontagu** [reply]
We are _definitely_ planning to release a package I think we will call `pydantic_next` for precisely this purpose. (The same as `pydantic2` in your description above, I believe.)

I don't believe we currently have a plan to release a `pydantic1` or similar (@samuelcolvin may be able to correct this), but if there is a need for it that `pydantic_next` doesn't solve, I think it can be considered as well.

**Reply by tlambert03** [reply]
thanks @dmontagu! in looking around a bit, I found https://github.com/pydantic/pydantic/issues/5465 which says that `pydantic_next` is "Pydantic V2 in a separate namespace".  Am I correct in assuming that `pydantic` on PyPI will also point to v2 as soon as it comes out of pre-release?

If that's the case, then the best that a v1 package can do is to "hope" that all of the other packages in the environment that want to use pydantic v2 have declared `pydantic_next` as a dependency, and _not_ `pydantic>=2.0` ... do I have that correct?  If so, doesn't that still leave us with a big potential problem?

(in other words, as long as the main name `pydantic` becomes v2, then without a `pydantic_last` type namespace, there's nothing a v1 package can do to prevent being broken eventually)

**Reply by caniko** [reply]
No, Pydantic V2 will eventually be the default for `pydantic`.

I think `pydantic_next` is for a smoother migration experience for those that are willing to accept at least a temporary upper bound versioning, and doesn't address the issue you are outlining @tlambert03.

It is safe to say that packages from both Pydantic and tiangolo use what you refer to as *true* semantic versioning; you will have to do your own research for whatever else you use.

You should constrain the version in your own packages for packages that don't have the upper bounds we are referring to, so `pydantic = "^1"` in your own projects should be enough:

```toml
package-with-pydantic-v2 = "*"
pydantic = "^1"
```

This will only install v1, and not v2. I expect that `package-with-pydantic-v2` also supports v1. You can also add `pydantic_next` here, and start migrating!

**Reply by jni** [reply]
> I expect that package-with-pydantic-v2 also supports v1. 

Sorry can you clarify here @caniko? Currently there is no way for a package to use both pydantic v2 and v1 together, so `package-with-pydantic-v2` will necessarily specify `pydantic>=2` and conflict with our `pydantic<2` constraint?

I think a concrete example is indeed the right way to clarify the issue here. Suppose I maintain `package_a` which uses pydantic, and depend on two other packages `package_b` and `package_c` that also use pydantic (it's a popular project!). Without `pydantic_old` or a long-lived `pydantic_next`, I cannot migrate to v2 until *both* `package_b` and `package_c` do so, because it will be impossible to install all three packages in the same environment.

Fine, you might say, dependencies form a directed acyclic graph and we just need to traverse that graph in order and update packages accordingly. But this might take a rather long time, during which packages downstream in the graph are unable to use newer versions of packages upstream. And for this reason packages upstream might be reluctant to update until they know that a substantial proportion of their user community is *also* ready to update — resulting in a stalemate and further prolonging the transition.

This more or less summarises what happened with the Python 2-3 transition, which took pretty much a decade and required a lot of community effort to build adapters that would look the same in both versions (e.g. `six`), and most code limiting itself to the subset of Python that worked in both 2 and 3 (ie unable to really take advantage of Python 3 features).

**Reply by caniko** [reply]
You have `pydantic` v1, the other package is locked to v2: You need to upgrade to v2.

You have v1, upstream has v1 or v2: You can use poetry to lock to v1.

There are multiple packages that use v1 and/or v2. There is no solution even if there was `pydantic_old` because those packages use a locked version of `pydantic`. Your best bet is to fork, and make an upgrade PR towards V2 for every V1 package if you can't wait. There just is no way around it.

**Reply by jni** [reply]
> There just is no way around it.

There is, that is precisely the argument: if pydantic were to release differently-named packages for each major version, packages could coexist that use either of those differently named packages. napari could continue to use `pydantic1` while new plugins popped up using `pydantic2`, and our community could migrate our codebase gradually, without having a compatibility break.

Without this "escape hatch", popular upstream packages are indeed forced to support either 1 or 2, and thus limited to the subset of pydantic that is the same between 1 and 2, just like with Python 2->3.

**Reply by tlambert03** [reply]
This! 👆

**Reply by caniko** [reply]
Let's not use fallacies, every popular package will eventually be on V2, otherwise it goes against the definition of popular.

You could use the "escape hatch" argument for every major release in every ecosystem, and you would be fine for making it. You are free to make this request or fork and create pydantic_v1 on PyPi yourself. Nobody is stopping you.

**Reply by tlambert03** [reply]
> Nobody is stopping you.

@samuelcolvin politely requested that we don't do that.  so, let's not :)

**Reply by caniko** [reply]
Just don't name it so that it is confused with `pydantic`.

You can name it `oldantic_v1`, making sure the name is OK with Samuel is probably the best course forward.

**Comment by samuelcolvin** [comment]
The idea of `pydantic_next` is that you can get started with pydantic V2 while still using v1 else where in your codebase, and critically while packages like fastapi are still using pydantic v1.

This wouldn't work if we released a pydantic v1 package.

**Comment by tlambert03** [comment]
Thanks all for your comments.  I do still feel like my point has been missed.  Im not worried about _my_ packages or _my_ migration plan or strategy.  I know I can use pinning strategies, and such and migrate when I desire (as long as no other package shows up in an environment with one of my libraries with an incompatible pin).  As a developer on projects (like napari/napari) that have many plugins and complicated environment solving, my concern was more that pydantic will soon be causing unsolvable environments  quite a lot.  And that if it notices that possibility, it could do something to ameliorate it.

Anyway, I respect that you've got lots of stuff going on at the moment. Thanks for the conversation.

**Reply by samuelcolvin** [reply]
Somewhat related, I do think most packages that depend pydantic heavily should have a constraint of pydantic<2 to avoid lots of problems when pydantic v2 is released.

Not everyone is careful about pinning dependencies.

**Reply by tlambert03** [reply]
Yes that will be my advice to our communities as well, and I think I will probably discourage other developers from specifying pydantic>=2.0 for quite a long time, preferring to use pydantic_next instead

(If everyone who upgrades only uses pydantic_next, we'll be fine, but not if anyone upgrades and specifies pydantic>=2.0)

**Reply by caniko** [reply]
Why would any maintainer of any popular package ever do `pydantic>=2.0` without making sure this is OK? That sounds very inconsiderate to me, at least not before fixing important bugs for the previous major that is compatible with V1.

**Opinion**
I don't think `pydantic_next` is a good idea for packages, but rather for projects with many models that need time to migrate their huge base. I wouldn't ship v1 and v2 together in a package, that will just make the UX terrible.

Just make a separate branch and migrate at your own pace to V2 while staying at V1 on the main. I guess you could use `pydantic_next` here, it is a development branch; go wild.

**Reply by tlambert03** [reply]
> I wouldn't ship v1 and v2 together in a package, that will just make the UX terrible.

Me neither.  again, I'm not referring to a single package or a single app. I'm referring to many separate libraries that may find themselves installed together into the same environment (but don't necessarily know ahead of time who they will be installed with).  I don't know how to say it another way 😅.  Thanks again for the discussion

**Reply by caniko** [reply]
#programming@academia

**Reply by Czaki** [reply]
> Why would any maintainer of any popular package ever do `pydantic>=2.0` without making sure this is OK? That sounds very inconsiderate to me, at least not before fixing important bugs for the previous major that is compatible with V1.

In the context of [napari](https://github.com/napari/napari) the problem may not be popular packages but some scientific, less popular project that is used by some plugin (plugin could be also a small project). 

The core problem is that pip does not always satisfy the constraints of already installed packages. 

If we could ensure that every package is installed in a single `pip` call then having two packages. One with `pydantic<2` and the second with `pydantic>=2` then the `pip` resolver will fail with the proper message. But when installed in the sequence of two calls then `pip` may install pydantic>=2.0 instead of fail.

**Reply by tlambert03** [reply]
**edit**  it looks like you just deleted your post.  So to clarify my post below, note that I was responding to a comment from @caniko that said:

> This can be solved by using poetry.

----


Sorry, I really hate to keep saying you're missing the point, but you keep proposing things that are not relevant to the problem we're describing.  But rather solutions that an individual application that has a single deployment step (like, say, a fast API app or something) might do... but will not work for what we're saying.   poetry will not help us here.

Here's a complete scenario (using napari as an example, since that's the ecosystem I'm particularly thinking about) that I'm trying to explain:

1. Some random end-user creates a virtual environment and runs `pip install napari`.  (note, _we_ don't have any control over how our end users are going to install napari, they might use pip, they might use poetry, they might use conda, it's not in our control)
2. note that napari depends on pydantic, as well as at least three other libraries that depend on pydantic including npe2, psygnal, app-model, etc...  Since all of these now need to coordinate their updates to pydantic v2, the authors of these packages all try to communicate and make sure that none of them pins to `pydantic>=2.0` (knowing it would make them incompatible with the rest).  So if they want to update, they strictly use `pydantic_next` until every other package has the chance to update as well.  (so far so good).
3. Random end-user later learns about some napari plugin (let's call it `some-plugin`) and they run `pip install some-plugin` in their environment.  (`some-plugin` doesn't know anything about the need to synchronize the update ... they just read about all the cool things that pydantic v2 has to offer, used it and pinned it in their dependencies as you're "supposed" to).   
4. Pip happily updates pydantic to v2 (since it wasn't installed in the same session).  Random end-user has now broken their environment, and neither napari nor any of the other libraries work.
5. Random end-user now comes to us and tells us that napari doesn't work, since they don't understand the subtleties here.  (yes, of course, we could argue that we just need to educate the entire python community how these things work, but we could also try to do something to avoid this situation).

It's simple to just kick the problem elsewhere and say "well random end-user should learn better habits for virtual environments"... but if we had `pydantic_old`... then at the very least,  napari, npe2, magicgui, psygnal, app-model, and packages that know they need some time to update could protect themselves from the situation described above.  It wouldn't matter if pydantic v2 got installed into those environments, since they would have a safe namespace that will always keep their current code working, no matter what version of pydantic is installed.

**Reply by tlambert03** [reply]
If you genuinely don't understand or agree with the problem, that is ok... but would you have any particular objections to us releasing our own `pydantic_old` on PyPI and conda-forge?

**Reply by caniko** [reply]
@tlambert03 please be nice, you are escalating. I deleted it because I figured I didn't have time to argue, bedtime ☺️

You have pydantic v1, the other package is locked to v2: You need to upgrade to v2.

You have v1 upstream has v1 or v2: You can use poetry to lock to v1.

There are multiple packages that use v1 and/or v2. There is no solution even if there was `pydantic_old` because those packages use a locked version of `pydantic`. Your best bet is to fork, and make an upgrade PR towards V2 for every V1 package if you can't wait. There just is no way around it.

@tlambert03 the scenario you are talking about is from the perspective of an enduser. You should focus on your side of things. I outlined the options available. You still don't use `pydantic_next` the right way in your scenario IMHO. I really don't see anyone using `pydantic_next` this way.

**Reply by tlambert03** [reply]
Sorry you felt I was escalating.  I do appreciate the discussion, just kinda felt like we were going in circles a bit since poetry is a solution that an application with a one time deployment might use, but doesn't solve the library case.  Hope my example clarified the case I'm talking about :)

**Reply by tlambert03** [reply]
> You should focus on your side of things. I outlined the options available.

This is where we disagree I guess.  Before starting this thread I spent a few days trying to make a couple packages support both pydantic v1 and v2.  Seeing that it would be extremely hard if at all possible, I realized that there's very little I can do on my side of things to prevent myself from being broken from the scenario I describe above (and I'm afraid the options you outline above are insufficient as well, let me know if you disagree).

It's easy enough to say "don't think about it from the end user situation", but I feel that the pydantic_old scenario really does provide at least a partial solution to the problem that I haven't seen in any of your proposals.  Many packages have done that sort of thing (and that's the reason it is done)

As always, thanks for your input!

**Reply by Czaki** [reply]
> You have v1 upstream has v1 or v2: You can use poetry to lock to v1.

It is not a problem for ane experienced python user to lock to pydantic<2. One could use poetry, the * second one could use `.env` with proper variables. The problem happens when your package users are scientists/data scientists/data analysist etc and your app/library is extensible by users using plugins. 

Who do not have deep Python knowledge and will open issues or request support in another way.

**Comment by tlambert03** [comment]
one last thing... I did find this comment from @samuelcolvin: https://github.com/pydantic/pydantic/discussions/4649#discussioncomment-3950305

> Instead of `pydantic_next`, it might be better to flip it around and have `pydantic_old` to be explicit to encourage newer pydantic to be the default.

which got a number of +1s ... so it does sound like there had been at least a little internal support for this idea.

**Reply by samuelcolvin** [reply]
I'm very opposed to this on reflection, for reasons explained below https://github.com/pydantic/pydantic/discussions/5402#discussioncomment-5761643 - this package would never be upgraded thus users would be more tempted to stick with using it indefinitely - long after it's no longer supported.

As mentioned below, I think it's a bad idea to release any other packages to PyPI, instead we should have instructions for installing `pydantic_next` from git.

**Reply by tlambert03** [reply]
thanks for your input!  understood, and response below: https://github.com/pydantic/pydantic/discussions/5402#discussioncomment-5761679

**Comment by samuelcolvin** [comment]
The way this was solved with python 2 vs 3, and the way I think @tiangolo is intending to solve this with fastapi is to support both pydantic v1 and v2 for a while.

This might seem like lots of work, but actually it's probably as simple as copying your code and having two modules - one for V1 and one for V2.

**Reply by Czaki** [reply]
And if versions 1 and 2 live in different namespaces it will be simpler to write such a backward compatibility layer.

**Reply by samuelcolvin** [reply]
Honestly this whole conversation, and the amount of confusion, makes me want to avoid releasing any other packages.

I think releasing with `pydantic1` or `pydantic2` (or both) might lead to people thinking those are the correct ways to install pydantic, and with `pydantic1`/`pydantic_old` there's a temptation that since newer versions are never released, that they're still fine to use, long after we stop supporting them.

For this reason I think it would be a **Bad idea** for the community to release any other pydantic packages.

I think even we (the pydantic organisation) probably shouldn't add any other pckages to PyPI, instead we should suggest users install `pydantic_next` via git to make it as clear as possible that it's a temporary hack, not a package/solution you should use long term.

**Reply by tlambert03** [reply]
> For this reason I think it would be a Bad idea for the community to release any other pydantic packages.

Thanks for weighing in @samuelcolvin ... I did very much want to hear your opinion on this, so I appreciate it.  I'm absolutely dedicated to finding a solution for our ecosytem that you and the pydantic team are also happy with, and so wouldn't put out anything you weren't at least ok with on some level.  Hear you loud and clear here 👍

We'll definitely continuing to work on this and think about it.  I'll continue working on some code to see if I can create a solution for our stuff that I feel is robust enough to the kinds of problems we've been describing.  Haven't quite heard anything that I feel like is a really satisfying approach, but haven't given up :)

thanks again

**Comment by caniko** [comment]
Signing of this thread, it has been a tremendous brain-exercise trying to emphasize the pain and suffering for those that are affected by this issue. I really wish there was a golden bullet for this issue, but I don't think there is... Not something that makes everybody happy!

I guess the silver-lining is that everyone here has a lot of opinions, and expertise that can be put to better use in their IDEs where they actually do the work to move over to the more safe and efficient version of Pydantic, which is V2. I know it is dissatisfying to be told you have to work a little extra for something you didn't perhaps ask for, but Pydantic is mostly used in FastAPI (correct me if I am wrong). We come second, I personally can't wait to use `computed_fields`.

Kind regards

**Reply by samuelcolvin** [reply]
> Pydantic is mostly used in FastAPI (correct me if I am wrong)

That's not correct, only about 25% of pydantic's use is with fastapi - that's what the downloads suggest.

By my calculation, roughly another 25% of usage is with other libraries - then 50% of usage is direct, with no library.
